src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java

```diff
@@ -168,7 +178,7 @@ public synchronized SpoolMessage addMessage(Publish request) throws InterruptedE
         SpoolMessage message = SpoolMessage.builder().id(id).request(request).build();
         addMessageToSpooler(id, message);
         queueOfMessageId.putLast(id);
-
+        qos0MessageCheckRequired.set(true);
```

Comment:
check for qos0 here too?

Reply:
do we also need to set in `addId()` too?

Reply:
We don't need the check here, as in this case we want to have it set to true for any new message added. 
Look at case number 3 above in my comment. In that condition, even after doing a check once, it's not guaranteed that there are no more QOS0 messages left in queue, we just stopped because storage issue is solved. Now, _any_ new message addition would change those storage conditions and that's why should set the flag to true again for any new message not just a QOS0 message.

The reason we can have that check in `persistentQueueSync` function is because we don't replace QOS0 messages using `removeMessagesWithQosZero`(The only place where flag could go back to false) while doing the queue sync. 

For the addId(), we don't need it there as well. As that is only called by mqttclient when we try to publish a message (do a `popId` first), but actually fail to publish it and add it to retries (do a `addId`). That's not a new message being added but an old message that was popped from spooler queue (and never actually removed from spooler) but never published and added back to queue.

Reply:
I might be missing something, but always setting the flag on addMessage wouldn't give performance benefits?  subsequent calls to addMessage would always perform qos0 checks

Reply:
At queue capacity, it will only be triggered at every addMessage if the previous Removal Procedure actually resulted in adding a new message. Otherwise, if it didn’t, the flag will remain false as we exit `addMessage` with SpoolerStorageException. Subsequent addMessage checks are not performed then.